core/executor/src/allocator.rs

```diff
@@ -200,45 +195,56 @@ impl FreeingBumpHeapAllocator {
 		Ok(res)
 	}
 
-	fn le_bytes_to_u32(arr: [u8; 4]) -> u32 {
-		u32::from_le_bytes(arr)
-	}
-
-	fn write_u32_into_le_bytes(bytes: u32, slice: &mut [u8]) {
-		slice[..4].copy_from_slice(&bytes.to_le_bytes());
-	}
-
 	fn get_item_size_from_index(index: usize) -> usize {
 		// we shift 1 by three places, since the first possible item size is 8
 		1 << 3 << index
 	}
 
-	fn get_heap_4bytes(&mut self, ptr: u32) -> Result<[u8; 4]> {
-		let mut arr = [0u8; 4];
-		self.heap.get_into(self.ptr_offset + ptr, &mut arr)?;
-		Ok(arr)
-	}
-
-	fn get_heap_byte(&mut self, ptr: u32) -> Result<u8> {
-		let mut arr = [0u8; 1];
-		self.heap.get_into(self.ptr_offset + ptr, &mut arr)?;
-		Ok(arr[0])
+	// Read a u64 from the heap.
+	//
+	// u64 is encoded in memory in little-endian form with all bits flipped. The bit flip serves to
+	// make the high order bits 1 instead of 0, minimizing the chances of reading a corrupted by
+	// valid-looking value.
+	fn get_heap_u64(&self, heap: &[u8], offset: u32) -> Result<u64> {
+		let range = self.heap_range(offset, 8, heap.len())
+			.ok_or_else(|| error("read out of heap bounds"))?;
+		let bytes = heap[range].try_into()
+			.expect("[u8] slice of length 8 must be convertible to [u8; 8]");
+		Ok(!u64::from_le_bytes(bytes))
```

Comment:
I've read the explaination but still don't get this mechanism, can you ELI5?

Reply:
I'm glad you asked because I would be perfectly happy getting rid of this. I'd really like @cmichi 's opinion here as well.

The intent is to sort of mimic the behavior of [this assertion](https://github.com/paritytech/substrate/pull/3610/files#diff-5da990ee099f29391d7bf93c20a0b25bL171) while being simpler to implement. In the current version of the code, when you write the list index, which should fit in 8 bits, you set the other 64 - 8 bits to all 1s. Then when you read the list index, it asserts they are all 1s. My assumption is that this is because stretches of 7 consecutive bytes of 1 bits are rare in the heap normally, so it acts as some extra detection mechanism against heap corruption. We should probably decide whether that is necessary. BUT assuming we want to keep something like that, I think the easier way to do it is to just write a 64-bit int, read the whole thing as a u64, then check that it's in the expected range. If any of the higher-order bits got flipped, we could detect that the list index is out of range and there is some sort of corruption. On the other hand, bits in the heap are probably more likely to be 0 than 1 simply because the linear memory is 0-initialized and 0 is a more common default. So we would rather these detection bits be 1. So instead of just writing the u64 in LE as one normally would, we can do a bitwise flip as part of the encoding. Then if you read it back and and any of the high order bits are 0, then they will be 1 after flipping and the list index will be out of range, which is detectable.

Reply:
> In the current version of the code, when you write the list index, which should fit in 8 bits, you set the other 64 - 8 bits to all 1s.

AFAIR the intention was to make sure pointers are always aligned to 8 bytes, hence the long 8*8 bytes prefix where only the first 8 bits are used. The bitflip-detection was no intention originally, thoug it doesn't sound bad to me. You suggestion of using a u64 sounds good to me, the other refactorings also make the code better readable :+1: .

Reply:
TBH, I think that some other pattern other than all 1s, would work better here.
If we are going to leave this approach, I'd love to see this explaination in the code.

Reply:
I clarified the comment a bit.

Reply:
Removed the bit flipping because its confusing and questionably effective.